blob/service.go

```diff
@@ -347,127 +353,193 @@ func (s *Service) retrieve(
+			// parse the blob
+			sharesParser.length = blobLen
+			_, isComplete := sharesParser.addShares(appBlobShares)
 			if !isComplete {
-				appShares = nil
-				break
+				return nil, nil, fmt.Errorf("expected the shares to construct a full blob")
 			}
-			// otherwise construct blob
 			blob, err := sharesParser.parse()
 			if err != nil {
 				return nil, nil, err
 			}
 
-			if sharesParser.verify(blob) {
-				return blob, &proofs, nil
-			}
-
-			index += len(appShares) - len(shrs)
-			appShares = shrs
-			sharesParser.reset()
+			// number of shares per EDS row
+			numberOfSharesPerEDSRow := squareSize * 2
+			// number of shares from square start to namespace start
+			sharesFromSquareStartToNsStart := inclusiveNamespaceStartRowIndex * numberOfSharesPerEDSRow
+			// number of rows from namespace start row to current row
+			rowsFromNsStartToCurrentRow := currentShareIndex / squareSize
+			// number of shares from namespace row start to current row
+			sharesFromNsStartToCurrentRow := rowsFromNsStartToCurrentRow * numberOfSharesPerEDSRow
+			// number of shares from the beginning of current row to current share
+			sharesFromCurrentRowStart := currentShareIndex % squareSize
+			// setting the index manually since we didn't use the parser.set() method
+			blob.index = sharesFromSquareStartToNsStart +
+				sharesFromNsStartToCurrentRow +
+				sharesFromCurrentRowStart
+
+			if blob.Namespace().Equals(namespace) && sharesParser.verify(blob) {
+				// now that we found the requested blob, we will create
+				// its inclusion proof.
+				inclusiveBlobStartRowIndex := blob.index / (squareSize * 2)
+				exclusiveBlobEndRowIndex := inclusiveNamespaceStartRowIndex + exclusiveEndShareIndex/squareSize
+				if (currentShareIndex+blobLen)%squareSize != 0 {
+					// if the row is not complete with the blob shares,
+					// then we increment the exclusive blob end row index
+					// so that it's exclusive.
+					exclusiveBlobEndRowIndex++
+				}
 
-			if !wasEmpty {
-				// remove proofs for prev rows if verified blob spans multiple rows
-				proofs = proofs[len(proofs)-1:]
-			}
-		}
+				// create the row roots to data root inclusion proof
+				rowProofs := proveRowRootsToDataRoot(
+					append(header.DAH.RowRoots, header.DAH.ColumnRoots...),
+					inclusiveBlobStartRowIndex,
+					exclusiveBlobEndRowIndex,
+				)
+				rowRoots := make([]bytes.HexBytes, exclusiveBlobEndRowIndex-inclusiveBlobStartRowIndex)
+				for index, rowRoot := range header.DAH.RowRoots[inclusiveBlobStartRowIndex:exclusiveBlobEndRowIndex] {
+					rowRoots[index] = rowRoot
+				}
 
-		rowIndex++
-		if sharesParser.isEmpty() {
-			proofs = nil
-		}
-	}
+				// create the share to row root proofs
+				shareToRowRootProofs, _, err := pkgproof.CreateShareToRowRootProofs(
+					squareSize,
+					//nolint:lll
+					edsShares[inclusiveBlobStartRowIndex-inclusiveNamespaceStartRowIndex:exclusiveBlobEndRowIndex-inclusiveNamespaceStartRowIndex],
```

Comment:
Is there a bug? lower bound has 2 vars with `inclusive` and upped bound has 1 var with `exclusive` and 1 with `inclusive`. Looks sus.

Can we split this into 3 lines, please?
```
start := inclusiveBlobStartRowIndex-inclusiveNamespaceStartRowIndex
end := exclusiveBlobEndRowIndex-inclusiveNamespaceStartRowIndex
...
edsShares[start : end]
```

Reply:
Also not symmetrical `...StartRow - ...StartRow` BUT `...EndRow - ...StartRow`. Which might be ok but sus 🤷

Reply:
No bug in indexes, right? Just asking for the confirmation.

Reply:
yes, no bug. We're shifting the indexes by `inclusiveNamespaceStartRowIndex` because the EDS shares start from `[inclusiveNamespaceStartRowIndex: exclusiveNamespaceEndRowIndex]`